cpp/src/iterator/iterator.cuh

```diff
@@ -0,0 +1,395 @@
+    >;
+
+    CUDA_HOST_DEVICE_CALLABLE
+    column_input_iterator(const T_column_input col, const Iterator &it) : super_t(it), colData(col){}
+
+    CUDA_HOST_DEVICE_CALLABLE
+    column_input_iterator(const T_column_input col) : super_t(Iterator{0}), colData(col){}
+
+    CUDA_HOST_DEVICE_CALLABLE
+    column_input_iterator(const column_input_iterator &other) : super_t(other.base()), colData(other.colData){}
+
+    // befriend thrust::iterator_core_access to allow it access to the private interface below
+    friend class thrust::iterator_core_access;
+
+  private:
+    const T_column_input colData;
+
+    // it is private because only thrust::iterator_core_access needs access to it
+    CUDA_HOST_DEVICE_CALLABLE
+    typename super_t::reference dereference() const
+    {
+      gdf_index_type id = *(this->base()); // base() returns base iterator: `Iterator`
+      return  Transformer{}.operator()(colData.at(id));
+    }
+};
+
+} // namespace detail
+
+// ---------------------------------------------------------------------------
+// helper functions to make iterator
+
+/** -------------------------------------------------------------------------*
+ * @brief helper function to make iterator with nulls
+ * Input iterator which can be used for cub and thrust.
+ *
+ * @tparam T_element The cudf data type of input array
+ * @tparam T_output  The cudf data type of output data or array
+ * @tparam Transformer
+ *                   Transforms pair(value, bool) into T_output form.
+ *                   The default is `scalar_cast_transformer<T_output, T_element>`
+ * @tparam Iterator_Index
+ *                   The base iterator which gives the index of array.
+ *                   The default is `thrust::counting_iterator`
+ *
+ * @param[in] data     The pointer of column data array
+ * @param[in] valid    The pointer of null bitmask of column
+ * @param[in] identity The identity value used when the mask value is false
+ * @param[in] it       The index iterator, `thrust::counting_iterator` by default
+ * -------------------------------------------------------------------------**/
+template <bool has_nulls, typename T_element, typename T_output = T_element,
+    typename Transformer = cudf::detail::scalar_cast_transformer<T_output, T_element>,
+    typename Iterator_Index=thrust::counting_iterator<gdf_index_type> >
+auto make_iterator(const T_element *data, const bit_mask::bit_mask_t *valid,
+    T_element identity, Iterator_Index const it = Iterator_Index(0))
+{
+    using T_colunn_input = cudf::detail::column_input<T_element, has_nulls>;
+    using T_iterator = cudf::detail::column_input_iterator<T_output, T_colunn_input, Transformer, Iterator_Index>;
+
+    // column_input constructor checks if valid is not nullptr when has_nulls = true
+    return T_iterator(T_colunn_input(data, valid, identity), it);
+}
+
+/** -------------------------------------------------------------------------*
+ *  @overload auto make_iterator(const T_element *data,
```

Comment:
I think overload documentation should say what this overload does (e.g. how it is different).